describeconfigs.go

```diff
@@ -0,0 +1,157 @@
+
+	// ErrorCode
+	ErrorCode int16
+
+	// ErrorMessage
+	ErrorMessage string
+
+	// ConfigEntries
+	ConfigEntries []DescribeConfigResponseConfigEntry
+}
+
+// DescribeConfigResponseConfigEntry
+type DescribeConfigResponseConfigEntry struct {
+	ConfigName          string
+	ConfigValue         string
+	ReadOnly            bool
+	IsDefault           bool
+	IsSensitive         bool
+	ConfigSynonyms      []DescribeConfigResponseConfigSynonym
+	ConfigType          int8
+	ConfigDocumentation string
+}
+
+// DescribeConfigResponseConfigSynonym
+type DescribeConfigResponseConfigSynonym struct {
+	ConfigName   string
+	ConfigValue  string
+	ConfigSource int8
+}
+
+// DescribeConfigs sends a config altering request to a kafka broker and returns the
+// response.
+func (c *Client) DescribeConfigs(ctx context.Context, req *DescribeConfigsRequest) (*DescribeConfigsResponse, error) {
+	resources := make([]describeconfigs.RequestResource, len(req.Resources))
+
+	for i, t := range req.Resources {
+		configNames := make([]string, len(t.ConfigNames))
+		for _, v := range t.ConfigNames {
+			configNames = append(configNames, v)
+		}
+		resources[i] = describeconfigs.RequestResource{
+			ResourceType: t.ResourceType,
+			ResourceName: t.ResourceName,
+			ConfigNames:  configNames,
+		}
+	}
+
+	m, err := c.roundTrip(ctx, req.Addr, &describeconfigs.Request{
+		Resources:            resources,
+		IncludeSynonyms:      req.IncludeSynonyms,
+		IncludeDocumentation: req.IncludeDocumentation,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("kafka.(*Client).DescribeConfigs: %w", err)
+	}
+
+	res := m.(*describeconfigs.Response)
+	ret := &DescribeConfigsResponse{
+		Throttle:  makeDuration(res.ThrottleTimeMs),
+		Resources: make([]DescribeConfigResponseResource, len(res.Resources)),
+	}
+
+	for _, t := range res.Resources {
+
+		configEntries := make([]DescribeConfigResponseConfigEntry, len(t.ConfigEntries))
+		for _, v := range t.ConfigEntries {
+
+			configSynonyms := make([]DescribeConfigResponseConfigSynonym, len(v.ConfigSynonyms))
+			for _, cs := range v.ConfigSynonyms {
+				configSynonyms = append(configSynonyms, DescribeConfigResponseConfigSynonym{
+					ConfigName:   cs.ConfigName,
+					ConfigValue:  cs.ConfigValue,
+					ConfigSource: cs.ConfigSource,
+				})
+			}
+
+			configEntries = append(configEntries, DescribeConfigResponseConfigEntry{
+				ConfigName:          v.ConfigName,
+				ConfigValue:         v.ConfigValue,
+				ReadOnly:            v.ReadOnly,
+				IsDefault:           v.IsDefault,
+				IsSensitive:         v.IsSensitive,
+				ConfigSynonyms:      configSynonyms,
+				ConfigType:          v.ConfigType,
+				ConfigDocumentation: v.ConfigDocumentation,
+			})
+		}
+
+		resource := DescribeConfigResponseResource{
+			ResourceType:  t.ResourceType,
+			ResourceName:  t.ResourceName,
+			ErrorCode:     t.ErrorCode,
+			ErrorMessage:  t.ErrorMessage,
+			ConfigEntries: configEntries,
+		}
+		ret.Resources = append(ret.Resources, resource)
+	}
```

Comment:
```suggestion
	for i, t := range res.Resources {

		configEntries := make([]DescribeConfigResponseConfigEntry, len(t.ConfigEntries))
		for j, v := range t.ConfigEntries {

			configSynonyms := make([]DescribeConfigResponseConfigSynonym, len(v.ConfigSynonyms))
			for k, cs := range v.ConfigSynonyms {
				configSynonyms[k] = DescribeConfigResponseConfigSynonym{
					ConfigName:   cs.ConfigName,
					ConfigValue:  cs.ConfigValue,
					ConfigSource: cs.ConfigSource,
				}
			}

			configEntries[j] = DescribeConfigResponseConfigEntry{
				ConfigName:          v.ConfigName,
				ConfigValue:         v.ConfigValue,
				ReadOnly:            v.ReadOnly,
				IsDefault:           v.IsDefault,
				IsSensitive:         v.IsSensitive,
				ConfigSynonyms:      configSynonyms,
				ConfigType:          v.ConfigType,
				ConfigDocumentation: v.ConfigDocumentation,
			}
		}

		ret.Resources[k] = DescribeConfigResponseResource{
			ResourceType:  t.ResourceType,
			ResourceName:  t.ResourceName,
			ErrorCode:     t.ErrorCode,
			ErrorMessage:  t.ErrorMessage,
			ConfigEntries: configEntries,
		}
	}
```
^ if you pre-allocate a slice you need to use indices

Reply:
If i understand right, if we preallocated slice size, then append command will not dynamic update slice size yet.

Reply:
the syntax used here pre-allocates with zero-value elements, I think this shows what's going on: https://play.golang.org/p/wvnolPMcM-a

Basically setting a `length` without `capacity` will zero initialize the entire `length`
```
make([]T, length, capacity)
```
https://golang.org/ref/spec#Slice_types

Reply:
Yes, you are right, i forgot add length in make.